src/parquet/bloom.h

```diff
@@ -0,0 +1,191 @@
+// under the License.
+
+#ifndef PARQUET_BLOOM_H
+#define PARQUET_BLOOM_H
+
+#include <set>
+#include <cstdint>
+
+#include "parquet/types.h"
+#include "parquet/util/memory.h"
+
+namespace parquet {
+class OutputStream;
+
+// Bloom Filter is a compact structure to indicate whether an item is not in set or
+// probably in set. Bloom class is underlying class of Bloom Filter which stores a
+// bit set represents elements set, hash strategy and bloom filter algorithm.
+
+// Bloom Filter algorithm is implemented using block Bloom filters from Putze et al.'s
+// "Cache-,Hash- and Space-Efficient Bloom Filters". The basic idea is to hash the
+// item to a tiny Bloom Filter which size fit a single cache line or smaller. This
+// implementation sets 8 bits in each tiny Bloom Filter. Tiny bloom filter are 32
+// bytes to take advantage of 32-bytes SIMD instruction.
+
+class Bloom final {
+ public:
+  // Hash strategy available for bloom filter.
+  enum HashStrategy {
+    MURMUR3_X64_128
+  };
+
+  // Bloom filter algorithm.
+  enum Algorithm {
+    BLOCK
+  };
+
+  /**
+   * Default false positive probability value use to calculate optimal number of bits
+   * used by bloom filter.
+   */
+  static constexpr double DEFAULT_FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  static constexpr int HEADER_SIZE = 12;
+
+  // Bytes in a tiny bloom filter block.
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
+
+  // Default seed for hash function, to keep cross compatibility the seed is same as
+  // value defined in parquet-mr.
+  static constexpr int DEFAULT_SEED = 104729;
+
+  // Default maximum bloom filter size (need to discuss)
+  static constexpr int DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES = 16 * 1024 * 1024;
+
+  // The block-based algorithm needs eight odd SALT values to calculate eight indexes
+  // of bit to set, one bit in each 32-bit word.
+  static constexpr uint32_t SALT[8] = { 0x47b6137bU, 0x44974d91U, 0x8824ad5bU,
+      0xa2b7289dU, 0x705495c7U, 0x2df1424bU, 0x9efc4947U, 0x5c6bfb31U };
+
+  typedef void (*HashFunc)(const void *key, int length, uint32_t seed, void* out);
+
+ public:
+  /// Constructor of bloom filter. The range of num_bytes should be within
+  /// [BYTES_PER_FILTER_BLOCK, DEFAULT_MAXIMUM_BLOOM_FILTER_BYTES], it will be
+  /// round up/down to lower/upper bound if num_bytes is out of range. It will use
+  /// murmur3_x64_128 as its default hash function and the block based algorithm.
+  /// @param num_bytes The number of bytes to store bloom filter bitset.
+  explicit Bloom(uint32_t num_bytes);
+
+  /// Construct the bloom filter with given bit set, it is used when reconstruct
+  /// bloom filter from parquet file. It use murmur3_x64_128 as its default hash
+  /// function and block-based algorithm as default algorithm.
+  /// @param bitset The given bitset to construct bloom filter.
+  /// @param len Length of bitset.
+  Bloom(const uint8_t* bitset, uint32_t len);
```

Comment:
This is a little tricky: If we are going to treat `bitset` as if it were an array of `uint32_t`s, we run into http://eel.is/c++draft/basic.lval#11

> If a program attempts to access the stored value of an object through a glvalue of other than one of the following types the behavior is undefined:
>  * the dynamic type of the object,
> ...
>  * a char, unsigned char, or std​::​byte type.

So we need `uint8_t` to be `char`, `unsigned char`, or `std::byte`. We'll need to check for alignment, as well, because compilers are allowed to assume that pointers to `uint32_t` have alignment that is greater than that of `uint8_t`.

Let's think about how this will get hooked into other libraries an ask @rdblue , who is more experienced in this code base, if we can put alignment requirements on the data we'll end up needing to use this with.

In an ideal world, I'd like it to be 64-byte aligned.

Reply:
Generally types alignment is equal to types size ([Wiki](https://en.wikipedia.org/wiki/Data_structure_alignment), this may be different among compilers), so I think the last sentence you mentioned should be 64-bits aligned, right? 

Therefore, uint32_t should have an four bytes alignment generally, consider we are restricting bitset length to [32bytes, 16Mbytes], and also length is power of 2, then the uint32_t* must not contains any padding, right?

BTW, as I known, uint8_t is typedef of unsigned char, am I wrong?

Reply:
I do really mean 64-byte. That is the cache line size in most modern x86-64 CPUs. Alignment is especially useful for SIMD operations, in which misaligned access can cause a segfault.

Reply:
OK,  Now I understood your concern.  SIMD operation should have some performance penalty when access unaligned data due to cross cache line access or cache miss (Some tests here: http://www.agner.org/optimize/blog/read.php?i=415#423), but I think it should not cause a segfault. 

How about set minimum bloom filter size to 64 bytes to avoid this?

Reply:
I should  have misunderstanding here cause I thought set to 64 bytes will cover this.

I will ask @rdblue to share his points on this.

Reply:
Objects can have alignment that is smaller than their size:

    struct Array { char payload[11]; };
    static_assert(alignof(Array) < sizeof(Array),"");

Reply:
I agree, the structure should have default alignment of its maximum member alignment. 

Any other thought or advice on this?

Reply:
For 64 bytes alignment, how about we allocate  (size+63) bytes, and use std::align to do a alignment. Or use boost::alignment:aligned_alloc.

Reply:
I got your point from your email reply. 

How about we check bitset address alignment and do copy if not?  And of course document alignment requirement and performance impact.  Can we do this firstly and consider BF client/consumer when doing read-side task?

Reply:
Do you mean this kind segfalut? https://stackoverflow.com/questions/25596379/simd-intrinsics-segmentation-fault

Reply:
I can do this. I don't understand the last sentence, do you mean I should add alignment check and static_assert code as well?

Reply:
Yes, add these in the test code. The return value you get from an allocatino should be aligned at at least alignof(uint32_t) and also uint8_t should be std::is_same to unsigned char.